axum-extra/src/extract/cached.rs

```diff
@@ -0,0 +1,232 @@
+///
+/// ```rust
+/// use axum_extra::extract::Cached;
+/// use axum::{
+///     async_trait,
+///     extract::{FromRequest, RequestParts},
+///     body::{self, BoxBody},
+///     response::IntoResponse,
+///     http::{StatusCode, Response},
+/// };
+///
+/// #[derive(Clone)]
+/// struct Session { /* ... */ }
+///
+/// #[async_trait]
+/// impl<B> FromRequest<B> for Session
+/// where
+///     B: Send,
+/// {
+///     type Rejection = (StatusCode, String);
+///
+///     async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+///         // load session...
+///         # unimplemented!()
+///     }
+/// }
+///
+/// struct CurrentUser { /* ... */ }
+///
+/// #[async_trait]
+/// impl<B> FromRequest<B> for CurrentUser
+/// where
+///     B: Send,
+/// {
+///     type Rejection = Response<BoxBody>;
+///
+///     async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+///         // loading a `CurrentUser` requires first loading the `Session`
+///         //
+///         // by using `Cached<Session>` we avoid extracting the session more than
+///         // once, in case other extractors for the same request also loads the session
+///         let session: Session = Cached::<Session>::from_request(req)
+///             .await
+///             .map_err(|err| err.into_response().map(body::boxed))?
+///             .0;
+///
+///         // load user from session...
+///         # unimplemented!()
+///     }
+/// }
+///
+/// // handler that extracts the current user and the session
+/// //
+/// // the session will only be loaded once, even though `CurrentUser`
+/// // also loads it
+/// async fn handler(
+///     current_user: CurrentUser,
+///     // we have to use `Cached<Session>` here otherwise the
+///     // cached session would not be used
+///     Cached(session): Cached<Session>,
+/// ) {
+///     // ...
+/// }
+/// ```
+///
+/// [request extensions]: http::Extensions
+#[derive(Debug, Clone, Default)]
+pub struct Cached<T>(pub T);
+
+#[async_trait]
+impl<B, T> FromRequest<B> for Cached<T>
+where
+    B: Send,
+    T: FromRequest<B> + Clone + Send + Sync + 'static,
+{
+    type Rejection = CachedRejection<T::Rejection>;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        match Extension::<T>::from_request(req).await {
+            Ok(value) => Ok(Self(value.0)),
+            Err(ExtensionRejection::ExtensionsAlreadyExtracted(err)) => {
+                Err(CachedRejection::ExtensionsAlreadyExtracted(err))
+            }
+            Err(_) => {
+                let value = T::from_request(req).await.map_err(CachedRejection::Inner)?;
+
+                req.extensions_mut()
+                    .ok_or_else(|| {
+                        CachedRejection::ExtensionsAlreadyExtracted(
+                            ExtensionsAlreadyExtracted::default(),
+                        )
+                    })?
+                    .insert(value.clone());
```

Comment:
How about using `CacheEntry<T>` (where `CacheEntry` is just a private newtype) instead of `T` here so you can have both a `Cached<T>` and a "regular" request extension `T` on the same request?

This would also stop accidentally relying on the cache having been propagated earlier in request processing by using `Extension<T>` over `Cached<T>` somewhere.

Reply:
Good idea!